src/main/java/uk/gov/justice/digital/delius/controller/CourtReportController.java

```diff
@@ -0,0 +1,113 @@
+    }
+
+    @RequestMapping(value = "offenders/offenderId/{offenderId}/courtReports/{courtReportId}", method = RequestMethod.GET)
+    @JwtValidation
+    public ResponseEntity<CourtReport> getOffenderCourtReportByOffenderIdAndCourtReportId(final @RequestHeader HttpHeaders httpHeaders,
+                                                                                 final @PathVariable("offenderId") Long offenderId,
+                                                                                 final @PathVariable Long courtReportId) {
+
+        log.info("Call to getOffenderCourtReportByOffenderIdAndCourtReportId");
+        return courtReportResponseEntityOf(Optional.of(offenderId), courtReportId);
+    }
+
+    @RequestMapping(value = "offenders/nomsNumber/{nomsNumber}/courtReports/{courtReportId}", method = RequestMethod.GET)
+    @JwtValidation
+    public ResponseEntity<CourtReport> getOffenderCourtReportByNomsNumberAndCourtReportId(final @RequestHeader HttpHeaders httpHeaders,
+                                                                                                     final @PathVariable("nomsNumber") String nomsNumber,
+                                                                                                     final @PathVariable Long courtReportId) {
+
+        log.info("Call to getOffenderCourtReportByNomsNumberAndCourtReportId");
+        return courtReportResponseEntityOf(offenderService.offenderIdOfNomsNumber(nomsNumber), courtReportId);
+    }
+
+    @RequestMapping(value = "offenders/crn/{crn}/courtReports/{courtReportId}", method = RequestMethod.GET)
+    @JwtValidation
+    public ResponseEntity<CourtReport> getOffenderCourtReportByCrnAndCourtReportId(final @RequestHeader HttpHeaders httpHeaders,
+                                                                                              final @PathVariable("crn") String crn,
+                                                                                              final @PathVariable Long courtReportId) {
+
+        log.info("Call to getOffenderCourtReportByCrnAndCourtReportId");
+        return courtReportResponseEntityOf(offenderService.offenderIdOfCrn(crn), courtReportId);
+    }
+
+    private ResponseEntity<List<CourtReport>> courtReportsResponseEntityOf(Optional<Long> maybeOffenderId) {
+        return maybeOffenderId
+            .map(offenderId -> new ResponseEntity<>(courtReportService.courtReportsFor(offenderId), OK))
+            .orElseGet(this::courtReportsNotFound);
+    }
+
+    private ResponseEntity<CourtReport> courtReportResponseEntityOf(Optional<Long> maybeOffenderId,  Long courtReportId) {
+        Optional<CourtReport> maybeCourtReport = maybeOffenderId.flatMap(offenderId -> courtReportService.courtReportFor(offenderId, courtReportId));
+        return maybeCourtReport.map(
+                courtReport -> new ResponseEntity<>(courtReport, OK)).orElse(courtReportNotFound());
```

Comment:
Any reason to use orElse here bit orElseGet above? Either is OK in this case but I've been bitten by NPE when using orElse in the past with more complex payloads

Reply:
the notFound() methods don't do anything that would be any more expensive than constructing a new lambda function so probably just not worth it. I would usually use orElseGet if it was something that would be destructive if we executed that code or very expensive.